gnmi_server/server.go

```diff
@@ -515,3 +527,76 @@ func (s *Server) Capabilities(ctx context.Context, req *gnmipb.CapabilityRequest
 		GNMIVersion:        "0.7.0",
 		Extension:          exts}, nil
 }
+
+type uint128 struct {
+	High uint64
+	Low  uint64
+	sync.Mutex
+}
+
+func (lh *uint128) Compare(rh *uint128) int {
+	lh.Lock()
+	defer lh.Unlock()
+	if rh == nil {
+		// For MA disabled case, EID supposed to be 0.
+		rh = &uint128{High: 0, Low: 0}
+	}
+	if lh.High > rh.High {
+		return 1
+	}
+	if lh.High < rh.High {
+		return -1
+	}
+	if lh.Low > rh.Low {
+		return 1
+	}
+	if lh.Low < rh.Low {
+		return -1
+	}
+	return 0
+}
+
+// ReqFromMasterEnabledMA returns true if the request is sent by the master
+// controller.
+func ReqFromMasterEnabledMA(req *gnmipb.SetRequest, masterEID *uint128) bool {
+	// Read the election_id.
+	reqEID := uint128{High: 0, Low: 0}
+	// It can be one of many extensions, so iterate through them to find it.
+	for _, e := range req.GetExtension() {
+		ma := e.GetMasterArbitration()
+		if ma == nil {
+			continue
+		}
+		// The Master Arbitration descriptor has been found.
+		if ma.ElectionId == nil {
+			log.V(1).Infoln("No election id provided by master.")
+			continue 
```

Comment:
Spec expects InvalidArgument error for this case, but I guess the code results in a PermissionDenied error. Better to change signature of ReqFromMasterEnabledMA function to return an error.

```go
func ReqFromMasterEnabledMA(req *gnmipb.SetRequest, masterEID *uint128) error {
    ....
    if ma.ElectionId == nil {
        return status.Errorf(codes.InvalidArgument, "MA: ElectionId missing")
    }
    if ma.Role != nil {
        return status.Errorf(codes.Unimplemented, "MA: Role is not implemented")
    }
    ....
    ...
    if request EID < master EID {
        return status.Errorf(codes.PermissionDenied, "Not a master with election ID %v", masterEID)
    }
}

func (s *Server) Set(ctx context.Context, req *gnmipb.SetRequest) (*gnmipb.SetResponse, error) {
    if err := s.ReqFromMaster(req, &s.masterEID); err != nil {
        return nil, err
    }
    ....
```

Reply:
Thanks. func has been modified.